src/main/java/org/xbill/DNS/Client.java

```diff
@@ -17,39 +18,47 @@
   /** Packet logger, if available. */
   private static PacketLogger packetLogger = null;
 
-  private static volatile boolean run;
+  private static volatile boolean run = true;
+  private static final List<Runnable> timeoutTasks = new CopyOnWriteArrayList<>();
+  private static final List<Runnable> closeTasks = new CopyOnWriteArrayList<>();
   private static Thread selectorThread;
-  private static List<Runnable> timeoutTasks = new CopyOnWriteArrayList<>();
-  static Selector selector;
+  private static volatile Selector selector;
 
   interface KeyProcessor {
     void processReadyKey(SelectionKey key);
   }
 
-  static void start() throws IOException {
-    if (run) {
-      return;
+  static Selector selector() throws IOException {
+    if (selector == null) {
+      synchronized (Client.class) {
+        if (selector == null) {
+          selector = Selector.open();
```

Comment:
> This isn't the usual pattern for if-null/synchronized/if-null, as it sets the visible static field before the synchronised section is complete. The variable that is used as the check should not be set until the last instruction in the if block in this pattern. In this case, it could cause a very rare null condition for the `selectorThread` static field which is referenced in the `close` method.

I might be missing something, but how? `close` can only be called _after_ selector is initialized. The worst that could happen here is an `IOException` when calling `open()`. But in this case neither the selector thread nor the shutdown hook will be registered, and everything is retried when either of the Nio[Tcp/Udp]Clients attempt to get the selector.
Also, having access to the selector before the selector thread is started or the shutdown hook registered does not matter, but is in fact this is wanted. Queuing of commands to process can and shall be done while the selector thread starts up.

> 
> A method I have used in the past for static initialisation is to use your previous `AtomicBoolean` approach to restrict entries (without global synchronisation in this step), and use a `CountDownLatch` with 1 token to prevent exits before the initialisation is complete. An example of that here could be:
> 
> ```
> private static final AtomicBoolean initialisationStarted - new AtomicBoolean(false);
> private static final CountDownLatch initialisationComplete = new CountDownLatch(1);
> 
> static Selector selector() throws IOException {
>   if(initialisationStarted.compareAndSet(false, true)) {
>     try {
>       selector = Selector.open();
>       log.debug("Starting dnsjava NIO selector thread");
>       selectorThread = new Thread(Client::runSelector);
>       selectorThread.setDaemon(true);
>       selectorThread.setName("dnsjava NIO selector");
>       selectorThread.start();
>       Thread closeThread = new Thread(Client::close);
>       closeThread.setName("dnsjava NIO shutdown hook");
>       Runtime.getRuntime().addShutdownHook(closeThread);
>     } finally {
>       initialisationCompleted.countDown();
>     }
>   }
>   initialisationCompleted.await();
>   return selector;
> }
> ```

If `selector.open()` fails with an `IOException`, there will never be a retry and further calls to the method will return null.